### PR TITLE
update sample code analyzer regexp & test case

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

PR #30129 failed to pass the sample code analyzer because [the analyzer miscategorized the sample code](https://github.com/flutter/flutter/pull/30129/checks?check_run_id=87681336). 

This pr updates the regex the analyzer uses, so it should be able to recognize expressions such as 
`const Foo a = Foo();` as "other code" rather than a constructor call.

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: dev/bots/analyze-sample-code.dart [hunk 1]
```diff
@@ -165,7 +165,7 @@ class SampleChecker {
   static final RegExp _codeBlockEndRegex = RegExp(r'///\s+```\s*$');
 
   /// A RegExp that matches a Dart constructor.
-  static final RegExp _constructorRegExp = RegExp(r'[A-Z][a-zA-Z0-9<>.]*\(');
+  static final RegExp _constructorRegExp = RegExp(r'(const )?[A-Z][a-zA-Z0-9<>.]*\(');
```
Comment (gspencergoog) [inline dev/bots/analyze-sample-code.dart]: Seems like it should handle private constructors too.
```suggestion
  static final RegExp _constructorRegExp = RegExp(r'(const\s+)?_*[A-Z][a-zA-Z0-9<>._]*\(');
```

File: dev/bots/test/analyze-sample-code-test-input/known_broken_documentation.dart [hunk 2]
```diff
@@ -41,3 +41,16 @@
...
+/// {@tool sample}
+/// ```dart
+/// const Text('Poor wandering ones!')
+/// ```
+/// {@end-tool}
+///
+/// {@tool sample}
+/// ```dart
+/// const text0 = Text('Poor wandering ones!');
+/// const text1 = Text('Poor wandering ones!');
```
Comment (gspencergoog) [inline dev/bots/test/analyze-sample-code-test-input/known_broken_documentation.dart]:
```suggestion
/// const text1 = Text('Poor wandering ones!');
/// const text2 = const   _Text('Poor wandering ones!');
```

File: dev/bots/test/analyze-sample-code-test-input/known_broken_documentation.dart [hunk 3]
```diff
@@ -41,3 +41,16 @@
 /// )
 /// ```
 /// {@end-tool}
+///
+/// {@tool sample}
+/// ```dart
+/// const Text('Poor wandering ones!')
```
Comment (gspencergoog) [inline dev/bots/test/analyze-sample-code-test-input/known_broken_documentation.dart]:
```suggestion
/// const Text('Poor wandering ones!')
/// const  _Text('Poor wandering ones!')
```

**Review by gspencergoog — CHANGES_REQUESTED** [review]

File: dev/bots/test/analyze-sample-code-test-input/known_broken_documentation.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply dev/bots/test/analyze-sample-code-test-input/known_broken_documentation.dart]: I was under the impression that `const text2 = const   _Text('Poor wandering ones!');` violates the style guide. Surprisingly the analyzer didn't complain.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Pull request merged**